src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java

```diff
@@ -55,16 +66,42 @@
     @CheckForNull
     private final String credentialsId;
 
+    /**
+     * Jenkins Server Root URL to be used by that Bitbucket endpoint.
+     * The global setting from Jenkins.getActiveInstance().getRootUrl()
+     * will be used if this field is null or equals an empty string.
+     * This variable is bound to the UI, so an empty value is saved
+     * and returned by getter as such.
+     */
+    private final String bitbucketJenkinsRootUrl;
+
+    /**
+     * Value of not-empty bitbucketJenkinsRootUrl normalized for end-users
+     * and saved, to avoid recalculating it over and over; an empty or null
+     * value still causes evaluation of current global setting every time.
+     */
+    private transient String endpointJenkinsRootUrl;
+
+
     /**
      * Constructor.
      *
      * @param manageHooks   {@code true} if and only if Jenkins is supposed to auto-manage hooks for this end-point.
      * @param credentialsId The {@link StandardCredentials#getId()} of the credentials to use for
      *                      auto-management of hooks.
+     * @param bitbucketJenkinsRootUrl The custom (or empty for global setting) Jenkins Root URL
+     *                      auto-management of hooks.
      */
-    AbstractBitbucketEndpoint(boolean manageHooks, @CheckForNull String credentialsId) {
+    AbstractBitbucketEndpoint(boolean manageHooks, @CheckForNull String credentialsId, @CheckForNull String bitbucketJenkinsRootUrl) {
```

Comment:
You are passing the objects in the wrong order. 
The abstractBitbucketEndpoint should own this Jenkins URL might differ per endpoint.
It should just call `this.getJenkinsRootUrl()` which is a method that either takes the custom Jenkins URL or the instance root URL. I would just skip the whole transient normalized thingy... and choose to store it on the bitbucketJenkinsRootUrl.

Reply:
I am not sure how to interpret this, could you please rephrase in layman terms? :-)
What would the correct order be?

As for the transient normalized, as I commented elsewhere, it is important that the URL string returned to callers conforms to some of their expectations. Normalization is a moderately costly but definitive translation from one string (user's input stored in config) to another string (normalized value), so I figured it made sense to only normalize once and not on every getter access.

Reply:
I don't think this should be passed to the constructor